### PR TITLE
Add support for host mode port PublishMode in services

### PR DESCRIPTION
**Pull request by mrjana**
Add api/cli support for adding host port PublishMode in services.

```
mrjana@dev-1:~/docker/scripts$ docker service create --name ss2 --publish-host 5000 --publish-ingress 5000 --network foo mrjana/simpleweb simpleweb

mrjana@dev-1:~/docker/scripts$ docker service ps ss2
NAME                             IMAGE             NODE     DESIRED STATE  CURRENT STATE          ERROR  PORTS
ss2.1.xlgwmnomhp609divvy8xdqh5w  mrjana/simpleweb  worker1  Running        Running 9 seconds ago         *:32768->5000/tcp
```
Signed-off-by: Jana Radhakrishnan <mrjana@docker.com>

### CODE REVIEW

File: cli/command/service/create.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ func newCreateCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.StringSliceVar(&opts.constraints, flagConstraint, []string{}, "Placement constraints")
 	flags.StringSliceVar(&opts.networks, flagNetwork, []string{}, "Network attachments")
 	flags.VarP(&opts.endpoint.ports, flagPublish, "p", "Publish a port as a node port")
```
Comment (mavenugo) [inline cli/command/service/create.go]: Given that we have 2 independent publish flags, should we move -p above to hidden ?
`flags.MarkHidden("p")`

File: cli/command/service/opts.go [hunk 2]
```diff
@@ -349,7 +351,31 @@ func (e *endpointOptions) ToEndpointSpec() *swarm.EndpointSpec {
 	ports, portBindings, _ := nat.ParsePortSpecs(e.ports.GetAll())
 
 	for port := range ports {
-		portConfigs = append(portConfigs, convertPortToPortConfig(port, portBindings)...)
+		rawPortConfigs := convertPortToPortConfig(port, portBindings)
```
Comment (mavenugo) [inline cli/command/service/opts.go]: should we instead change `convertPortToPortConfig` to take in PublishMode instead of walking the configs again as follows ?

File: cli/command/service/update.go [hunk 3]
```diff
@@ -46,14 +46,18 @@ func newUpdateCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.Var(newListOptsVar(), flagLabelRemove, "Remove a label by its key")
 	flags.Var(newListOptsVar(), flagContainerLabelRemove, "Remove a container label by its key")
 	flags.Var(newListOptsVar(), flagMountRemove, "Remove a mount by its target path")
-	flags.Var(newListOptsVar(), flagPublishRemove, "Remove a published port by its target port")
+	flags.Var(newListOptsVar(), flagPublishRemove, "Remove a published ingress port by its target port")
```
Comment (mavenugo) [inline cli/command/service/update.go]: Similar to the above comment, shall we hide the existing flag ?

**Review by mavenugo — COMMENTED** [review]
@mrjana Can you pls document if this will impact existing services (without the PublishMode) and also any compatibility concerns with a mixed cluster of 1.13 manager and 1.12 based workers and vice-versa. Also things to consider for upgrades and downgrades.

**Comment by mavenugo** [comment]
The design was reviewed under https://github.com/docker/swarmkit/pull/1645 . moving to code-review.

**Comment by thaJeztah** [comment]
I put it back to design, I'm a bit confused; we now have _three_ flags (6 if we look at `docker service update`) to expose a port.

I can see that being very confusing for users, and wonder if we can find a better approach

**Comment by mavenugo** [comment]
@thaJeztah ok. do you have any suggestions ? Lets start from there.

**Comment by thaJeztah** [comment]
Current options in `--publish` on `docker run` (not `service create`) to see what
things we _may_ expect

description | example
---|---
publish a container's port to a random port on the host | `-p 80`
publish a container's port to a specific port on the host | `-p 1234:80`
publish a container's port to a random host port within a specified range | `-p 8000-9000:80`
publish a range of container ports to random ports on the host | `-p 8000-8010`
publish a range of container ports to a range of host ports | `-p 8000-8010:9000-9010`
specify the protocol to publish (udp, tcp) | `-p 80/udp`, `-p 1234:80/udp`, `-p 8000-9000:80/udp`, `-p 8000-8010/udp`, `-p 8000-8010:9000-9010/udp`
publish the port only on specific IP-addresses | `-p 127.0.0.1:90:80 -p 192.168.202.120:90:80`
publish a container port on a random port on a specific IP-address | `-p 127.0.0.1::80` (double colon means value omitted)
publish a container port to multiple random ports on a specific IP-address | `-p 127.0.0.1::80 -p 127.0.0.1::80`

**Comment by thaJeztah** [comment]
Some feature requests;

- named ports (https://github.com/docker/docker/issues/21322)
- Assigning service published ports to IP (https://github.com/docker/docker/issues/26696)
- restrict port access to specific IP-addresses (https://github.com/docker/docker/issues/14105)
- Support multiple network interfaces in Swarm (https://github.com/docker/docker/issues/24317)

**Comment by thaJeztah** [comment]
So, my concerns are that (first of all) we tried the `-p` / `--publish` flag to be consistent with `docker run` so that it's easier for people to find themselve "at home" when using `docker service create`. By "deprecating" / renaming the `--publish` flag, that's going to be confusing.

Also, more features will be added in future, and we end up with even more flags. I'm not sure if the current `-p` syntax will make that easy to use. My initial thought was to use a syntax similar to `--mount` for publishing, as it's extendable, and non ambiguous (albeit verbose), something like

    --new-publish-flag target-port=80,published-port=8080,mode=host

In future we can add "porcelain" flags for easier use, if requested.

**Comment by mavenugo** [comment]
@thaJeztah I like this idea which helps with the issue of "grouping" related flags together at the CLI level (which is ofcourse possible in engine-API & protobuf structures). I wish we make this a consistent pattern and introduce such a "grouping" concept that can be used for more such use-cases.

**Comment by yank1** [comment]
LGTM

**Comment by mavenugo** [comment]
@icecrime @cpuguy83 @aluzzardi can you please add your comments / opinion on the UX as proposed by this PR vs @thaJeztah's proposal ?

Given the code-freeze is round the corner, it will be great to have an answer soon so that we can also do integration-testing with WS2016 as well.

**Comment by cpuguy83** [comment]
The `--mount` style syntax is ok for me.

**Comment by icecrime** [comment]
Sebastiaan's proposal makes sense to me. Yes it's verbose, but it's maybe better to get the fundamentals right (as in: non-ambiguous, and future-proof).

**Comment by vdemeester** [comment]
Same as @cpuguy83 and @icecrime, I think it makes sense 👼

**Comment by mavenugo** [comment]
thanks @icecrime @cpuguy83 @vdemeester.

@thaJeztah @mrjana shall we use --publish flag itself and overload it with both the formats ?
I think the text parsing is mutually exclusive and we already have sane defaults for the older format.

For example,

`-p 80` -> `-p target-port=80,mode=ingress`
`-p 1234:80` -> `-p target-port=80,published-port=1234,mode=ingress`

etc...

**Comment by thaJeztah** [comment]
If we can do it "technically", I am not against reusing the flag

One thing we should look at is how "removing" a published port will work in `docker service update` (see also https://github.com/docker/docker/pull/25860)

**Comment by mrjana** [comment]
ping @dnephin. Can you please chime in on the port publication format that is being discussed here? This will override the micro format specification that has been used so far.

File: cli/command/service/opts.go [hunk 4]
```diff
@@ -644,8 +672,14 @@ const (
 	flagName                  = "name"
 	flagNetwork               = "network"
 	flagPublish               = "publish"
+	flagPublishIngress        = "publish-ingress"
+	flagPublishHost           = "publish-host"
 	flagPublishRemove         = "publish-rm"
+	flagPublishRemoveIngress  = "publish-rm-ingress"
+	flagPublishRemoveHost     = "publish-rm-host"
```
Comment (dnephin) [inline cli/command/service/opts.go]: These should be `publish-ingress-rm`, `publish-host-rm`, same with the add flags

File: cli/command/service/create.go (hunk 1 above)
Comment (dnephin) [inline cli/command/service/create.go]: Do we need all 3 of these?

Doesn't the old flag do `ingress`? So can't we keep the existing flag for ingress and add only `publish-host` ?

**Comment by dnephin** [comment]
Oh, I see the proposal for having the new flags use a different format. I think that makes sense. We need a way to expose more information, and following the mount syntax sounds correct.

I think we should deprecate the old `--publush` flag at the same time. Deprecating it will hide it, but still allow it to be used for now.

**Comment by mrjana** [comment]
@dnephin Are you suggesting adding a new flag for the new format and deprecating and hiding the `-publish` flag? If so, what would be your suggestion for the new flag name?

**Comment by mavenugo** [comment]
@mrjana @dnephin `--port` ?

**Comment by dnephin** [comment]
Yes, that's my suggestion. A new flag that uses a CSV format (like `--mount`) which will allow is to include all the details (name, mode, etc).

`--port` sounds appropriate to me.

**Comment by mrjana** [comment]
PR has been updated based on the new format for port options. PTAL

File: cli/command/service/create.go [hunk 5]
```diff
@@ -40,10 +40,12 @@ func newCreateCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.Var(&opts.constraints, flagConstraint, "Placement constraints")
 	flags.Var(&opts.networks, flagNetwork, "Network attachments")
 	flags.VarP(&opts.endpoint.ports, flagPublish, "p", "Publish a port as a node port")
+	flags.MarkHidden(flagPublish)
```
Comment (cpuguy83) [inline cli/command/service/create.go]: Do we really need to hide it vs just specify it's for ingress mode?

File: cli/command/service/opts.go [hunk 6]
```diff
@@ -553,4 +620,9 @@ const (
 	flagHealthRetries         = "health-retries"
 	flagHealthTimeout         = "health-timeout"
 	flagNoHealthcheck         = "no-healthcheck"
+
+	portOptTargetPort    = "target-port"
```
Comment (cpuguy83) [inline cli/command/service/opts.go]: `-port` on these are redundant.

File: daemon/cluster/executor/container/controller.go [hunk 7]
```diff
@@ -424,6 +440,64 @@ func parseContainerStatus(ctnr types.ContainerJSON) (*api.ContainerStatus, error
...
+
+		protocol := api.ProtocolTCP
+		switch strings.ToLower(parts[1]) {
+		case "tcp":
+			protocol = api.ProtocolTCP
+		case "udp":
+			protocol = api.ProtocolUDP
+		default:
+			return nil, fmt.Errorf("invalid protocol: %s", parts[1])
+		}
```
Comment (dnephin) [inline daemon/cluster/executor/container/controller.go]: I think we already have code to do this in `go-connections/nat`

File: cli/command/service/update.go [hunk 8]
```diff
@@ -627,14 +665,46 @@ portLoop:
 				continue portLoop
 			}
 		}
+
+		for _, portCSV := range removePorts {
+			portKVPairs, err := parsePortCSV(portCSV)
```
Comment (dnephin) [inline cli/command/service/update.go]: `removePorts` should be just the container port numbers / protocol, not the entire list of values pairs.

That's the behaviour we have for other fields.

File: cli/command/service/opts.go [hunk 9]
```diff
@@ -291,8 +308,55 @@ func (opts *healthCheckOptions) toHealthConfig() (*container.HealthConfig, error
 	return healthConfig, nil
 }
 
+func parsePortCSV(value string) (map[string]string, error) {
+	kvPair := make(map[string]string)
```
Comment (dnephin) [inline cli/command/service/opts.go]: Please make this a flag type like `MountOpt`. It will make the code below a lot cleaner when you go to read the values.

It also means that we validate during flag parsing instead of waiting until converting the types.

File: cli/command/service/update.go [hunk 10]
```diff
@@ -588,26 +593,58 @@ func portConfigToString(portConfig *swarm.PortConfig) string {
...
+				pPort, _ := strconv.ParseUint(portKVPairs[portOptPublishedPort], 10, 16)
+				newConfigs = append(newConfigs, swarm.PortConfig{
+					Protocol:      swarm.PortConfigProtocol(portKVPairs[portOptProtocol]),
+					TargetPort:    uint32(tPort),
+					PublishedPort: uint32(pPort),
+					PublishMode:   swarm.PortConfigPublishMode(portKVPairs[portOptMode]),
+				})
+
+				continue
+			}
```
Comment (dnephin) [inline cli/command/service/update.go]: This code will be much cleaner when we have a `PortOpt` type (as requested above).

File: cli/command/service/create.go (hunk 5 above)
Comment (dnephin) [inline cli/command/service/create.go]: I think in the latest PR `--port` replaces `--publish`, so we would need to hide it.

Comment (cpuguy83) [inline cli/command/service/create.go]: It's not really replacing it, just providing an alternative/more versatile flag.
But.. ¯\_(ツ)_/¯

Comment (thaJeztah) [inline cli/command/service/create.go]: I'm okay with keeping it visible, it's similar to `--volume` vs `--mount` on `docker run`, but happy to hear if we feel strongly about hiding it

File: cli/command/service/update.go (hunk 8 above)
Comment (mrjana) [inline cli/command/service/update.go]: It is not the entire list. Only `target-port` is mandatory. But the user can be more specific if they want to.

File: cli/command/service/create.go (hunk 5 above)
Comment (mrjana) [inline cli/command/service/create.go]: Do we need keep it hidden or visible? I am fine either way. Please let me know.

File: cli/command/service/opts.go (hunk 6 above)
Comment (mrjana) [inline cli/command/service/opts.go]: Will remove.

**Comment by mrjana** [comment]
@dnephin I do agree that introducing `PortOpt` type would be better but it should be done to handle both the old microformat and the new expanded port format. Looking at it, it is going to result in a bit of refactoring. I think we should take it up as a separate PR.

File: cli/command/task/print.go [hunk 11]
```diff
@@ -51,7 +66,7 @@ func Print(dockerCli *command.DockerCli, ctx context.Context, tasks []swarm.Task
 
 	// Ignore flushing errors
 	defer writer.Flush()
-	fmt.Fprintln(writer, strings.Join([]string{"NAME", "IMAGE", "NODE", "DESIRED STATE", "CURRENT STATE", "ERROR"}, "\t"))
+	fmt.Fprintln(writer, strings.Join([]string{"NAME", "IMAGE", "NODE", "DESIRED STATE", "CURRENT STATE", "ERROR", "PORTS"}, "\t"))
```
Comment (thaJeztah) [inline cli/command/task/print.go]: The `PORTS` column is only propagated if `mode=host`, correct?

Also, we may want to consider swapping `PORTS` and `ERROR`, but we can had that discussion in a follow up.

**Review by thaJeztah — COMMENTED** [review]
Ok, so right now both of these work:

``` bash
docker service create --port 8080:80 --name web nginx:alpine

docker service create --publish 8080:80 --name web nginx:alpine
```

And both of these;

``` bash
docker service create \
  --publish mode=host,target=80,published=8080,protocol=tcp \
  --name web \
  nginx:alpine

docker service create \
  --port mode=host,target=80,published=8080,protocol=tcp \
  --name web \
  nginx:alpine
```

Since we decided to add a _new_ flag for this feature, instead of reusing the
existing `--publish` flag, I'm wondering if we should support both formats on
both flags. I think we should have;
1. `--publish` / `-p` only support the "old" microformat, and only support `ingress` publishing ("current behavior")
2. `--port` only support the "csv" format, OR
3. `--port` to support both formats

Users can use the `--publish` flag for most common things, and switch to `--port`
for advanced features.

I'm still doubting between `2.` and `3.`; if we want to keep the old "microformat",
and officially deprecate `--publish`, it should be `3.`, but this may be confusing,
because the microformat doesn't support the new options, and likely never will,
so it may be odd to add a "deprecated" format to a "new" flag.

File: cli/command/task/print.go (hunk 11 above)
Comment (thaJeztah) [inline cli/command/task/print.go]: There was some discussion about this (i.e. remove `COMMAND` (done)), and add `PORTS` instead, but perhaps best to make this change a separate PR

**Comment by vdemeester** [comment]
So if we were to start *from scratch*, I think we would choose `--publish` as a flag so, what I would propose is the following :

- Have `--publish` supports both format for now, *but* deprecate the old format for this flag (making `--publish` the *porcelain* flag, as `--mount` is)
- Have a new `--port` flag that support the old format and is seen as more *high-level*.

(I also agree with @cpuguy83 on the `ports` in the default format :stuck_out_tongue:)

**Comment by mavenugo** [comment]
@vdemeester i brought up the idea of supporting both formats under `--publish`. But @dnephin suggested otherwise https://github.com/docker/docker/pull/27917#issuecomment-259238370.
Can you both please come to a consensus on which is better for the user ? From a backward compatibility and ease of use for the users, I personally favor supporting both formats.
If we go down this path, then adding a new `--port` flag doesn't make much sense.

But if we decide to stick to `--port`, it has its advantages as well. First of all, I think this represents the concept better than `--publish` as we can expand the scope of the format in the future to use ports and not necessarily for publishing. We could have other modes that are purely internal usage within the cluster. It is hypothetical, but that is one of the benefits of `--port`.

But i dont agree with 
>> Have a new --port flag that support the old format and is seen as more high-level.

**Comment by vdemeester** [comment]
> 
But if we decide to stick to `--port`, it has its advantages as well. First of all, I think this represents the concept better than `--publish` as we can expand the scope of the format in the future to use ports and not necessarily for publishing. We could have other modes that are purely internal usage within the cluster. It is hypothetical, but that is one of the benefits of `--port`.

Fair enough. Then Let's deprecate `--publish` (and probably not supporting new format) and have `--port` supporting only the new format. @dnephin sounds good to you ? 👼

**Comment by dnephin** [comment]
SGTM

**Comment by vdemeester** [comment]
Moving to code-review then :wink:

**Comment by thaJeztah** [comment]
Ok, so `1.` and `2.` from my list;

1. `--publish` / `-p` only support the "old" microformat, and only support ingress publishing ("current behavior")
2. `--port` only supports the "csv" format

Thanks!

**Comment by mrjana** [comment]
Ok updated based on the latest agreement. 
Also @dnephin since we decided to not accept old format in `--port` and new format in `--publish` I re-implemented the `--port` part with  new `PortOpt` type. PTAL.

**Comment by mavenugo** [comment]
@mrjana the CI failures seems genuine

```
20:37:59 --- FAIL: TestUpdatePortsDuplicateKeys (0.00s)
20:37:59 	assert.go:84: update_test.go:249: Expected '1' (int) got '2' (int)
20:37:59 --- FAIL: TestUpdatePortsConflictingFlags (0.00s)
20:37:59 	assert.go:84: update_test.go:264: Expected an error, but error was nil
20:37:59 FAIL
```

**Comment by mrjana** [comment]
@mavenugo fixed

File: cli/command/service/update.go [hunk 12]
```diff
@@ -641,6 +651,15 @@ func updatePorts(flags *pflag.FlagSet, portConfig *[]swarm.PortConfig) error {
 		}
 	}
 
+	if flags.Changed(flagPortAdd) {
+		for _, entry := range flags.Lookup(flagPortAdd).Value.(*opts.PortOpt).Value() {
+			if v, ok := portSet[portConfigToString(&entry)]; ok && v != entry {
+				return fmt.Errorf("conflicting port mapping between %v:%v/%s and %v:%v/%s", entry.PublishedPort, entry.TargetPort, entry.Protocol, v.PublishedPort, v.TargetPort, v.Protocol)
```
Comment (dnephin) [inline cli/command/service/update.go]: I personally would prefer that only one of `--port` or `--published` should be used per command. Mixing them is confusing.

File: daemon/cluster/executor/container/controller.go (hunk 7 above)
Comment (mrjana) [inline daemon/cluster/executor/container/controller.go]: I don't think there is anything from go-connections/nat that can reused here to convert portMap to `api.PortConfig`.

File: cli/command/service/update.go (hunk 12 above)
Comment (mrjana) [inline cli/command/service/update.go]: This is as good as mixing any two different options. Mostly the users are going to start using the new style and forget about the old style since it is going to be deprecated. So I don't think the additional complexity is worth it.

Comment (mavenugo) [inline cli/command/service/update.go]: Yes. additional complexity isn't worth it here.

**Comment by mavenugo** [comment]
@mrjana can you please rebase ? :(

**Comment by vieux** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM

**Comment by mrjana** [comment]
Rebased

**Comment by thaJeztah** [comment]
Linking https://github.com/docker/docker/pull/28463#issuecomment-260940492 which should be documented as part of the documentation changes for this PR

**Comment by rootsongjc** [comment]
@mavenugo @thaJeztah  I use this command it works.
` docker service create --port mode=host,target=80,published=8080,protocol=tcp nginx:1.9
`
How many mode can I use?
I want to use my own bridge network "mynet", how can I specify it within service?

**Comment by thaJeztah** [comment]
@rootsongjc no, currently only "host" or "ingress".

Be aware that after UX reviews, the `--port` and `--publish` flags will be combined in 1.13 GA, see the discussion on https://github.com/docker/docker/issues/28527#issuecomment-262817805

**Comment by stevvooe** [comment]
Here was the original UX proposal:

https://github.com/docker/swarmkit/pull/1645#issuecomment-254684412

For mount, the syntax complexity made sense due the use of volume options and arbitrary filesystem paths. Ports are much, much simpler and forcing people to navigate this morass for simple port definition isn't great.

I am not sure I understood the main impetus behind using the csv-style syntax.

**Comment by Richard-Mathie** [comment]
Oh no, this seems to make things more complex.

does this mean there is no straight --publish [port:]port syntax defaulting to the ingress network any more? Is there any documentation on how this affects DNS resolution and any differences between networking for the two modes. I guess this helps to fix external connections to services like kafka, and removes a potential routing hop if you are using nginx.

Is that the rational behind this?

I also dont see the documentation for this in https://github.com/docker/docker/blob/master/docs/reference/commandline/service_create.md
what was the syntax you eventually went with?

**Comment by thaJeztah** [comment]
@Richard-Mathie no worries, the "simple" syntax will stay as well; see https://github.com/docker/docker/pull/28943, and the discussion on https://github.com/docker/docker/issues/28527

**Comment by thaJeztah** [comment]
Documentation is added here; https://github.com/mstanleyjones/docker.github.io/blob/2248f78dabc2b0e5092fd550ad0aa3c15d41e2b0/engine/swarm/services.md, but we should add it to the command line reference as well
